app.py

```diff
@@ -112,21 +111,90 @@ def get_bundles(entry):
     return bundles
 
 
-@app.route("/donate")
+def do_charge_or_show_errors(**kwargs):
+    app.logger.debug("----Retrieving Stripe customer...")
+
+    try:
+        email = request.form["stripeEmail"]
+        installment_period = request.form["installment_period"]
+        amount = request.form["amount"]
+        customer = stripe.Customer.create(
+            email=email, card=request.form["stripeToken"]
+        )
+        add_donation.delay(customer=customer, form=clean(request.form))
+
+        gtm = {
+            "event_value": amount,
+            "event_label": "once" if installment_period == "None" else installment_period
+        }
+
+        return render_template("charge.html",
+            gtm=gtm,
+            bundles=get_bundles("charge")
+        )
+    except stripe.error.CardError as e:
+        body = e.json_body
+        err = body.get("error", {})
+        message = err.get("message", "")
+        form_data = request.form.to_dict()
+        del form_data["stripeToken"]
+
+        return render_template(kwargs["template"],
+            bundles=kwargs["bundles"],
+            key=app.config["STRIPE_KEYS"]["publishable_key"],
+            message=message,
+            form_data=form_data
+        )
+
+
+def validate_form(FormType, **kwargs):
+    app.logger.info(request.form)
+
+    form = FormType(request.form)
+    email = request.form["stripeEmail"]
+
+    if not validate_email(email):
+        message = "There was an issue saving your email address."
+        return render_template("error.html", message=message)
+    if not form.validate():
+        app.logger.warning(f"Form validation errors: {form.errors}")
+        message = "There was an issue saving your donation information."
+        return render_template("error.html", message=message)
+
+    return do_charge_or_show_errors(
+        bundles=kwargs["bundles"],
+        template=kwargs["template"]
+    )
+
+
+@app.route("/donate", methods=["GET", "POST"])
 def member2_form():
     bundles = get_bundles("donate")
-    return render_template(
-        "member-form2.html",
+    template = "member-form2.html"
+
+    if request.method == "POST":
+        return validate_form(DonateForm,
+            bundles=bundles,
+            template=template
+        )
+
+    return render_template(template,
         bundles=bundles,
         key=app.config["STRIPE_KEYS"]["publishable_key"],
     )
 
-
-@app.route("/circleform")
+@app.route("/circleform", methods=["GET", "POST"])
```

Comment:
Maybe not in this PR, but I'd be in favor of making this consistent by dropping the `form` and renaming it to `circle` (and redirecting) like `donate`.

Reply:
I agree. I'll do that separately.